src/inmanta/server/services/resourceservice.py

```diff
@@ -414,6 +414,109 @@ async def mark_deployed(
+            async with connection.transaction():
+                # validate resources
+                resource = await data.Resource.get_one(
+                    environment=env.id,
+                    resource_id=resource_id,
+                    model=version,
+                    # acquire lock on Resource before read and before lock on ResourceAction to prevent conflicts with
+                    # cascading deletes
+                    lock=data.RowLockMode.FOR_UPDATE,
+                    connection=connection,
+                )
+                if not resource:
+                    raise NotFound("The resource with the given ids do not exist in the given environment.")
+
+                # no escape from terminal
+                if resource.status != status and resource.status in TERMINAL_STATES:
+                    if not fail_on_error:
+                        return
+                    else:
+                        LOGGER.error("Attempting to set undeployable resource to deployable state")
+                        raise AssertionError("Attempting to set undeployable resource to deployable state")
+
+                resource_action = data.ResourceAction(
+                    environment=env.id,
+                    version=version,
+                    resource_version_ids=[resource_version_id],
+                    action_id=uuid.uuid4(),
+                    action=const.ResourceAction.deploy,
+                    started=timestamp,
+                    messages=[
+                        {
+                            **logline.dict(),
+                            "timestamp": logline.timestamp.astimezone().isoformat(timespec="microseconds"),
+                        }
+                    ],
+                    changes={},
+                    status=status,
+                    change=const.Change.nochange,
+                    finished=timestamp,
+                )
+                await resource_action.insert(connection=connection)
+
+                self.log_resource_action(
+                    env.id,
+                    [resource_version_id],
+                    logline.level.to_int,
+                    logline.timestamp,
+                    logline.msg,
+                )
+
+                self.clear_env_cache(env)
+
+                extra_fields = {}
+                if status == ResourceState.deployed and not is_increment_notification:
+                    extra_fields["last_success"] = timestamp
+
+                await resource.update_fields(
+                    last_deploy=timestamp,
+                    status=status,
+                    last_non_deploying_status=const.NonDeployingResourceState(status),
+                    **extra_fields,
+                    connection=connection,
+                )
+
+                if "purged" in resource.attributes and resource.attributes["purged"] and status == const.ResourceState.deployed:
```

Comment:
perhaps I should remove this as well

Reply:
I think my other comment covers this. Ping me if you want more input.